internal/controller/clientpool/clientpool.go

```diff
@@ -131,8 +172,33 @@ func (cp *ClientPool) Close() {
 	defer cp.mux.Unlock()
 
 	for _, c := range cp.clients {
-		c.Close()
+		c.Client.Close()
+	}
+
+	cp.clients = make(map[ClientPoolKey]ClientInfo)
+}
+
+func (cp *ClientPool) isCertificateExpiredSoon(certBytes []byte, howSoon time.Duration) (bool, error) {
+	if len(certBytes) == 0 {
+		return false, nil
+	}
+
+	block, _ := pem.Decode(certBytes)
+	if block == nil {
+		return true, fmt.Errorf("failed to decode PEM block")
+	}
+
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return true, fmt.Errorf("failed to parse certificate: %v", err)
+	}
+
+	expiryTime := cert.NotAfter.Add(-howSoon)
+
+	// Check if certificate is expired or will expire within the safety buffer
+	if time.Now().After(expiryTime) {
```

Comment:
nit: I feel like the way you're setting up the `expiryTime` variable is a bit hard to read, because the real expiry time is `cert.NotAfter`, so it's confusing to subtract from that time and still call it `expiryTime`. Does that make sense?

Suggestion:

```suggestion
	almostExpiredTime := cert.NotAfter.Add(-howSoon)

	// Check if certificate is expired or will expire within the safety buffer
	if time.Now().After(almostExpiredTime) {
```